libraries/AP_Airspeed/AP_Airspeed_Health.cpp

```diff
@@ -0,0 +1,75 @@
+#include <AP_Common/AP_Common.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Math/AP_Math.h>
+#include "AP_Airspeed.h"
+
+bool AP_Airspeed::check_airspeed_health(uint8_t i)
+{
+    bool airspeed_status=true;
+
+    if((AP_AIRSPEED_FAILURE_MASK_BASIC & _options) && airspeed_status && (AP::ahrs().get_max_wind() > 0)) {
+        airspeed_status = check_airspeed_health_basic(i);
+    }
+
+    return airspeed_status;
+}
+
+bool AP_Airspeed::check_airspeed_health_basic(uint8_t i)
+{
+    float aspeed = get_airspeed();
+    bool airspeed_status=true;
+
+
+    float wind_max = AP::ahrs().get_max_wind();
+    uint32_t current_time = AP_HAL::millis();
+    uint32_t time_since_last_call =  current_time - state[i].airspeed_failure_last_check;
+    if(time_since_last_call > AP_AIRSPEED_FAILURE_TIME_PERIOD_MS) {
+        state[i].airspeed_failure_last_check = current_time;
+
+        if(param[i].use != 0) {
+            if (wind_max > 0 && AP::gps().status() >= AP_GPS::GPS_OK_FIX_2D) {
+                float gnd_speed = AP::gps().ground_speed();
+
+                // return false as the airspeed is not as expected and could be faulty
+                if(aspeed > (gnd_speed + wind_max) || aspeed < (gnd_speed - wind_max)) {
+                    state[i].airspeed_failure_basic_failed_count++;
+
+                    if(state[i].airspeed_failure_basic_failed_count > AP_AIRSPEED_FAILURE_NUM_FAULTS ) {
+                        gcs().send_text(MAV_SEVERITY_WARNING, "Airspeed fault-AHRS_WIND_MAX exceeded - %d disabled", i+1);
+                        state[i].reenable_use = param[i].use;
+                        param[i].use.set_and_notify(0);
+                        state[i].airspeed_failure_basic_renable = true;
+                        airspeed_status = false;
+                        state[i].healthy = false;
+                    }
+                }
+            }
+        } else if (state[i].airspeed_failure_basic_renable && (AP_AIRSPEED_FAILURE_MASK_BASICENABLE & _options)) {
+            if (wind_max > 0 && AP::gps().status() >= AP_GPS::GPS_OK_FIX_2D) {
+                float gnd_speed = AP::gps().ground_speed();
+
+                if(aspeed < (gnd_speed + wind_max) && aspeed > (gnd_speed - wind_max)) {
+                    state[i].airspeed_failure_basic_success_count++;
```

Comment:
Again, the count is related to sample rate. What you're actually looking for is one second of consistently good samples and any bad sample will reset the one second timer. So, timestamp the last bad sample and if you ever see a bad sample set it to now_ms and if (now_ms - last_bad_ms < 1000) then you know you're in the bad state. Right now yo have timers setting counters. Just use timers!

Reply:
Are you referring to just the enabling portion?  If so i can agree with that but over a longer period.

If your also referring to the initiating the failure part i disagree.  There are times where it drops say as a turn is made into the wind, or short gust catches it.  If you have an alternative where i can do that without counters of some kind that would be great.  A single bad reading shouldnt cause a failure.

Reply:
I'll use the low pass filter option, much better solution